src/estimagic/parameters/nonlinear_constraints.py

```diff
@@ -0,0 +1,468 @@
+        estimagic's reparametrization features. In this function we compute the Jacobian
+        of the constraint at the internal parameters using information on the Jacobian
+        of the constraint at the selected external parameters.
+
+        """
+        params = converter.params_from_internal(x)
+        select = external_selector(params)
+        jac = jacobian(select)
+        jac_matrix = block_tree_to_matrix(jac, constraint_eval, selected)
+        jac_sparse = _extend_jacobian(jac_matrix, selection_indices, n_params)
+        jac_internal = converter.derivative_to_internal(jac_sparse, x, jac_is_flat=True)
+        return np.atleast_2d(jac_internal)
+
+    # ==================================================================================
+    # Transform constraint function and derive bounds
+    # ==================================================================================
+    _type = "eq" if "value" in c else "ineq"
+
+    if _type == "eq":
+
+        # ==============================================================================
+        # Equality constraints
+        #
+        # We define the internal constraint function to be satisfied if it is equal
+        # to zero, by subtracting the fixed value.
+
+        _value = np.atleast_1d(np.array(c["value"], dtype=float))
+
+        def constraint_from_internal(x):
+            params = converter.params_from_internal(x)
+            select = external_selector(params)
+            out = np.atleast_1d(constraint_func(select)) - _value
+            return out
+
+        jacobian_from_internal = _jacobian_from_internal
+        n_constr = _n_constr
+
+    else:
+
+        # ==============================================================================
+        # Inequality constraints
+        #
+        # We define the internal constraint function to be satisfied if it is
+        # greater than or equal to zero (positivity constraint). If the bounds already
+        # satify this condition we do not change anything, otherwise we need to perform
+        # a transformation.
+
+        def _constraint_from_internal(x):
+            params = converter.params_from_internal(x)
+            select = external_selector(params)
+            return np.atleast_1d(constraint_func(select))
+
+        lower_bounds = c.get("lower_bounds", 0)
+        upper_bounds = c.get("upper_bounds", np.inf)
+
+        transformation = _get_transformation(lower_bounds, upper_bounds)
+
+        constraint_from_internal = _compose_funcs(
+            _constraint_from_internal, transformation["func"]
+        )
+
+        jacobian_from_internal = _compose_funcs(
+            _jacobian_from_internal, transformation["derivative"]
+        )
+
+        n_constr = 2 * _n_constr if transformation["name"] == "stack" else _n_constr
+
+    internal_constr = {
+        "n_constr": n_constr,
+        "type": _type,
+        "fun": constraint_from_internal,  # internal name for 'func'
```

Comment:
Can we use `func` instead of `fun`?

Reply:
For the internal interface I chose ``fun`` because most of the optimizers that can handle nonlinear constraints, expect the argument ``fun``.